MERGE OWNERSHIP AND MUTE (CHAT):

AdrianWeaver opened the pull request:
Adding chat features, such as transfer, promote, demote and mute.
Reworking guards for Owner and Admins.

Fixed a few mistakes including a 500 error on a missing catch.

Allowing numbers in commit title (githook).

Close #38 
Fix #51
Fix #53 
(#51 je t'aimeuhj'enboiraisdestonneaux)

riblanc commented on back/src/chat/channel/channel.controller.ts:
```diff
@@ -34,62 +37,118 @@ export class ChannelController {
 
-	@Post(':id/invite/:userid')
+	@Post(':targetChannelId/invite/:targetUserId')
 	@UseGuards(...AuthenticatedGuard)
```
Can't you use `isAdminGuard` on this route?
Thus avoiding the check on the [channelService](https://github.com/42rave/transcendence/blob/7ca3341ad5fa4dfac5c404c8d338d7a26e106c8a/back/src/chat/channel/channel.service.ts#L308-L312)

AdrianWeaver replied:
Indeed but I think I liked the idea of being able to tell that the reason you cannot invite someone in the channel is because the channel does not exist, not because you are not admin.
What do you think? Maybe @Elena-Lou or @Deacllock could share their thoughts?

Quick recap:
To invite someone in the channel (add them on the invite list) you need to be admin.
There is a guard (IsAdminGuard) that checks this condition and returns true or false therefore allowing or rejecting the request on the route if the conditions are met/not met.

Using this guard here would still mean one database query, less code inside the function (so cleaner code) but would not differentiate if the guard action comes from a non existing channel or from a user not admin of said channel.

riblanc replied:
You can throw an error on the guard:

```ts
async canActivate(context: ExecutionContext): Promise<boolean> {
  [...]
  if (!await this.channelService.isUserAdminInChannel(targetUserId, targetChannelId))
    throw new ForbiddenException('You are not admin');
  return true;
}
```

The default guard behavior is:
- If the canActivate return true nothing happens
- else it throw something like BadRequest or Forbidden exception (don't remember exactly sorry).

But you can if you want throw a custom error depending on a condition, else return true!

AdrianWeaver replied:
Ok gotcha.
But then I end up doing two requests in the guard for :
- "does the channel exist ?" 
- "is the user admin ?"

riblanc replied:
Yeah i guess..
Perhaps the best thing to do would be to implement middleware to retrieve the targetChannelId if given in the parameters, and inject it into the request context, so that the guards only have to execute an if statement instead of fetching the channel from the database twice.
But honestly, fetching the channel twice is acceptable, as long as the use of routes using these guards isn't really prone to spam.

riblanc replied:
I guess, fetching the channel twice it's less effort than implementing a new middleware etc.

riblanc reviewed (commented):
Honestly, it's becoming more and more enjoyable to read your code and we can feel that you're already much more at ease, what a pleasure!

AdrianWeaver commented:
I fixed several issues.
I made better use of the guards. Though there might be some work to be done on the kick and ban methods.
I fixed the join method with protected channel as requested by @Elena-Lou, now users should be able to join channels when they are not the owners and have the proper password :)

Elena-Lou reviewed (approved):
thank you for your work and the quick fix !
